orttraining/orttraining/core/graph/optimizer_builder.h

```diff
@@ -122,6 +127,49 @@ class OptimizerBuilder {
       std::vector<ArgDef>& output_weight_argdefs,
       std::vector<ArgDef>& output_gradient_argdefs,
       const bool enable_grad_clipping) const = 0;
+  
+    /**
+   * Adds the optimizer node to the graph.
+   * This component may be placed into an If node subgraph to enable
+   * conditional execution.
+   *
+   * @param weight_argdef The ArgDef of the weight to optimize.
+   * @param gradient_argdef The ArgDef of the gradient of the weight to
+            optimize.
+   * @param enable_grad_clipping The flag to force gradient clipping. If planning
+   *        to use the default behavior of each sub-class, use the other Build()
+   *        function without this argument.
+   * @param gradient_norm_argdef (Optional) The ArgDef of gradient norm.
+   * @param gradient_norm_finite_argdef (Optional) The ArgDef indicates whether
+            the passed-in gradient norm is finite.
+   * @param shared_optim_state (Optional) The initial state for optimizer params
+   *        shared by all weights.
+   * @param opt_config The optimizer configuration.
+   * @param[out] graph_defs The GraphDefs corresponding to the graph (possibly
+   *             a subgraph) that the component is to be added to.
+   * @param[out] new_external_initializers Any initializers that should be
+   *             placed in the parent graph, if there is one.
+   *             Other initializers are treated as local to the current
+   *             (sub)graph.
+   * @param[out] output_weight_argdefs The output weight ArgDef. All optimizers
+                 should have this output.
+   * @param[out] output_gradient_argdefs The output gradient ArgDef. All optimizers
+                 should have this output.
+   *
+   * @return The status of the optimizer node addition.
+   */
+  virtual Status Build(
```

Comment:
instead of adding another virtual Build() function with one extra parameter, it may be worthwhile to refactor Build() to take a struct (e.g., BuildConfig). less boilerplate to update if we need to extend it again. to handle the "default" behavior of enable_grad_clipping, it could be an optional<bool> with the empty state meaning the default

Reply:
great that we have just one Build() now! thanks for refactoring.
i see that an OptimizerGraphConfig parameter was added. that struct happens to have a bunch of other stuff too.
what i meant was maybe we can put all of the existing input parameters of Build()
```
      const std::vector<ArgDef>& weight_argdefs,
      const std::vector<ArgDef>& gradient_argdefs,
      const ArgDef* gradient_norm_argdef,
      const ArgDef* gradient_norm_finite_argdef,
      const std::vector<OptimizerNodeConfig>& opt_configs,
```
and enable_grad_norm_clip and shared_optimizer_states into a config struct just for Build().